### PR TITLE
Authentication Policies Settings Resource and Data Source

### PR DESCRIPTION
**Pull request by PingDavidR**
- closes PDI-1377

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,12 @@
+# v0.5.0 (unreleased)
+### FEATURES
+#### Resources
+* **New Resource:** `pingfederate_authentication_policies_settings` ([#150](https://github.com/pingidentity/terraform-provider-pingfederate/pull/150))
+
+#### Data Sources
+* **New Data Source:** `pingfederate_authentication_policies_settings` ([#T150](https://github.com/pingidentity/terraform-provider-pingfederate/pull/150))
```
Comment (henryrecker-pingidentity) [inline CHANGELOG.md:7]: Misplaced T here

File: examples/resources/pingfederate_authentication_policies_settings/import.sh [hunk 2]
```diff
@@ -0,0 +1,2 @@
+# "authenticationPoliciesSettingsId" should be "id"
```
Comment (henryrecker-pingidentity) [inline examples/resources/pingfederate_authentication_policies_settings/import.sh:1]: This is probably best for another issue, and I know this is copied from our other examples, but this comment doesn't seem accurate. In fact with resources like this that authenticationPoliciesSettingsId value shouldn't even matter, no matter what the user supplies.

File: internal/acctest/authenticationpoliciessettings/authentication_policies_settings_test.go [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+	}
+
+	resource.Test(t, resource.TestCase{
+		PreCheck: func() { acctest.ConfigurationPreCheck(t) },
+		ProtoV6ProviderFactories: map[string]func() (tfprotov6.ProviderServer, error){
+			"pingfederate": providerserver.NewProtocol6WithError(provider.NewTestProvider()),
+		},
+		Steps: []resource.TestStep{
+			{
+				Config: testAccAuthenticationPoliciesSettings(resourceName, initialResourceModel),
```
Comment (henryrecker-pingidentity) [inline internal/acctest/authenticationpoliciessettings/authentication_policies_settings_test.go:39]: To align with our other tests, the first step should set only the required properties of the resource. In this case there are no required properties, so the first step can just apply some HCL with no attributes in the resource itself. 
There should also be a 4th test step that returns to the initial empty model.

File: internal/resource/config/authenticationpolicies/settings/authentication_policies_settings_common.go [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+	EnableSpAuthnSelection  types.Bool   `tfsdk:"enable_sp_authn_selection"`
+}
+
+func readAuthenticationPoliciesSettingsResponse(ctx context.Context, r *client.AuthenticationPoliciesSettings, state *authenticationPoliciesSettingsModel, existingId *string) {
+	if existingId != nil {
+		state.Id = types.StringValue(*existingId)
+	} else {
+		state.Id = id.GenerateUUIDToState(existingId)
+	}
+	state.EnableIdpAuthnSelection = types.BoolValue(*r.EnableIdpAuthnSelection)
```
Comment (henryrecker-pingidentity) [inline internal/resource/config/authenticationpolicies/settings/authentication_policies_settings_common.go:23]: Use types.BoolPointerValue here and on the next line, rather than dereferencing the pointer

File: internal/resource/config/authenticationpolicies/settings/authentication_policies_settings_data_source.go [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+	providerConfig internaltypes.ProviderConfiguration
+	apiClient      *client.APIClient
+}
+
+// GetSchema defines the schema for the datasource.
+func (r *authenticationPoliciesSettingsDataSource) Schema(ctx context.Context, req datasource.SchemaRequest, resp *datasource.SchemaResponse) {
+	schema := schema.Schema{
+		Description: "Describes the authentication policies settings",
+		Attributes: map[string]schema.Attribute{
+			"id": schema.StringAttribute{
```
Comment (henryrecker-pingidentity) [inline internal/resource/config/authenticationpolicies/settings/authentication_policies_settings_data_source.go:37]: The id.ToDataSourceSchema call below adds `id` to the schema, so we don't need to duplicate it here

File: internal/resource/config/authenticationpolicies/settings/authentication_policies_settings_common.go [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+	"github.com/pingidentity/terraform-provider-pingfederate/internal/resource/common/id"
+)
+
+type authenticationPoliciesSettingsModel struct {
+	Id                      types.String `tfsdk:"id"`
+	EnableIdpAuthnSelection types.Bool   `tfsdk:"enable_idp_authn_selection"`
+	EnableSpAuthnSelection  types.Bool   `tfsdk:"enable_sp_authn_selection"`
+}
+
+func readAuthenticationPoliciesSettingsResponse(ctx context.Context, r *client.AuthenticationPoliciesSettings, state *authenticationPoliciesSettingsModel, existingId *string) {
```
Comment (henryrecker-pingidentity) [inline internal/resource/config/authenticationpolicies/settings/authentication_policies_settings_common.go:17]: Looks like `ctx` isn't used, so you can remove it from the parameters

File: internal/resource/config/authenticationpolicies/settings/authentication_policies_settings_resource.go [hunk 7]
```diff
@@ -0,0 +1,227 @@
...
+	providerConfig internaltypes.ProviderConfiguration
+	apiClient      *client.APIClient
+}
+
+// GetSchema defines the schema for the resource.
+func (r *authenticationPoliciesSettingsResource) Schema(ctx context.Context, req resource.SchemaRequest, resp *resource.SchemaResponse) {
+	schema := schema.Schema{
+		Description: "Manages Authentication Policies Settings",
+		Attributes: map[string]schema.Attribute{
+			"id": schema.StringAttribute{
```
Comment (henryrecker-pingidentity) [inline internal/resource/config/authenticationpolicies/settings/authentication_policies_settings_resource.go:40]: Same as the data source - no need to redefine id here

File: internal/resource/config/authenticationpolicies/settings/authentication_policies_settings_resource.go [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+		},
+	}
+
+	id.ToSchema(&schema)
+	resp.Schema = schema
+}
+
+func addOptionalAuthenticationPoliciesSettingsFields(ctx context.Context, addRequest *client.AuthenticationPoliciesSettings, plan authenticationPoliciesSettingsModel) error {
+
+	if internaltypes.IsDefined(plan.EnableIdpAuthnSelection) {
```
Comment (henryrecker-pingidentity) [inline internal/resource/config/authenticationpolicies/settings/authentication_policies_settings_resource.go:66]: No need for the IsDefined checks here - ValueBoolPointer already will return the correct values if one of these isn't defined (and they should be defined anyways since they both have a default value).

File: internal/resource/config/authenticationpolicies/settings/authentication_policies_settings_resource.go [hunk 9]
```diff
@@ -0,0 +1,227 @@
...
+				Default:     booldefault.StaticBool(false),
+			},
+		},
+	}
+
+	id.ToSchema(&schema)
+	resp.Schema = schema
+}
+
+func addOptionalAuthenticationPoliciesSettingsFields(ctx context.Context, addRequest *client.AuthenticationPoliciesSettings, plan authenticationPoliciesSettingsModel) error {
```
Comment (henryrecker-pingidentity) [inline internal/resource/config/authenticationpolicies/settings/authentication_policies_settings_resource.go:64]: No need to return an error from this method since it only will ever return nil. And `ctx` isn't used so it can be removed.

File: internal/resource/config/authenticationpolicies/settings/authentication_policies_settings_resource.go [hunk 10]
```diff
@@ -0,0 +1,227 @@
...
+		return
+	}
+
+	createAuthenticationPoliciesSettings := client.NewAuthenticationPoliciesSettings()
+	err := addOptionalAuthenticationPoliciesSettingsFields(ctx, createAuthenticationPoliciesSettings, plan)
+	if err != nil {
+		resp.Diagnostics.AddError("Failed to add optional properties to add request for authentication policies settings", err.Error())
+		return
+	}
+	requestJson, err := createAuthenticationPoliciesSettings.MarshalJSON()
```
Comment (henryrecker-pingidentity) [inline internal/resource/config/authenticationpolicies/settings/authentication_policies_settings_resource.go:109]: We decided to remove all these marshalJsons from the code so this should be removed along with any other marshalljson calls. If we're going to continue to scaffold resources with the code generator we need to update it to leave this out.

File: internal/resource/config/authenticationpolicies/settings/authentication_policies_settings_resource.go [hunk 11]
```diff
@@ -0,0 +1,227 @@
...
+	}
+	apiReadAuthenticationPoliciesSettings, httpResp, err := r.apiClient.AuthenticationPoliciesAPI.GetAuthenticationPolicySettings(config.ProviderBasicAuthContext(ctx, r.providerConfig)).Execute()
+
+	if err != nil {
+		if httpResp != nil && httpResp.StatusCode == 404 {
+			config.ReportHttpErrorAsWarning(ctx, &resp.Diagnostics, "An error occurred while getting the authentication policies settings", err, httpResp)
+			resp.State.RemoveResource(ctx)
+		} else {
+			config.ReportHttpError(ctx, &resp.Diagnostics, "An error occurred while getting the authentication policies settings", err, httpResp)
+		}
```
Comment (henryrecker-pingidentity) [inline internal/resource/config/authenticationpolicies/settings/authentication_policies_settings_resource.go:150]: Need a return after this line. Otherwise it will try to read into state from an empty response after a 404

File: internal/resource/config/authenticationpolicies/settings/authentication_policies_settings_resource.go [hunk 12]
```diff
@@ -0,0 +1,227 @@
...
+		resp.Diagnostics.AddError("Failed to add optional properties to add request for authentication policies settings", err.Error())
+		return
+	}
+	requestJson, err := createUpdateRequest.MarshalJSON()
+	if err == nil {
+		tflog.Debug(ctx, "Update request: "+string(requestJson))
+	}
+	updateAuthenticationPoliciesSettings = updateAuthenticationPoliciesSettings.Body(*createUpdateRequest)
+	updateAuthenticationPoliciesSettingsResponse, httpResp, err := r.apiClient.AuthenticationPoliciesAPI.UpdateAuthenticationPolicySettingsExecute(updateAuthenticationPoliciesSettings)
+	if err != nil && (httpResp == nil || httpResp.StatusCode != 404) {
```
Comment (henryrecker-pingidentity) [inline internal/resource/config/authenticationpolicies/settings/authentication_policies_settings_resource.go:197]: Update doesn't need 404 logic like this

**Review by henryrecker-pingidentity — CHANGES_REQUESTED** [review]

File: examples/resources/pingfederate_authentication_policies_settings/import.sh (hunk 2 above)
Comment (PingDavidR) [reply examples/resources/pingfederate_authentication_policies_settings/import.sh:1]: OK, I'll leave it here for consistency and we can do a global fix after researching this out.

File: internal/resource/config/authenticationpolicies/settings/authentication_policies_settings_common.go (hunk 4 above)
Comment (PingDavidR) [reply internal/resource/config/authenticationpolicies/settings/authentication_policies_settings_common.go:23]: Done

File: internal/resource/config/authenticationpolicies/settings/authentication_policies_settings_resource.go (hunk 10 above)
Comment (henryrecker-pingidentity) [reply internal/resource/config/authenticationpolicies/settings/authentication_policies_settings_resource.go:91]: Still need to remove these

File: internal/resource/config/authenticationpolicies/settings/authentication_policies_settings_common.go (hunk 6 above)
Comment (henryrecker-pingidentity) [reply internal/resource/config/authenticationpolicies/settings/authentication_policies_settings_common.go:17]: Still need to remove ctx from this method

File: internal/acctest/authenticationpoliciessettings/authentication_policies_settings_test.go (hunk 3 above)
Comment (henryrecker-pingidentity) [reply internal/acctest/authenticationpoliciessettings/authentication_policies_settings_test.go:46]: There should be four steps. Minimal -> Maximal -> Import -> back to Minimal.

File: internal/acctest/authenticationpoliciessettings/authentication_policies_settings_test.go [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+// Attributes to test with. Add optional properties to test here if desired.
+type authenticationPoliciesSettingsResourceModel struct {
+	enableIdpAuthnSelection bool
+	enableSpAuthnSelection  bool
+}
+
+func TestAccAuthenticationPoliciesSettings(t *testing.T) {
+	resourceName := "myAuthenticationPoliciesSettings"
+	// No required resources, test empty resource
+	emptyResourceModel := authenticationPoliciesSettingsResourceModel{}
```
Comment (henryrecker-pingidentity) [inline internal/acctest/authenticationpoliciessettings/authentication_policies_settings_test.go:24]: This won't result in actual empty HCL being sent, because you are still setting the attributes in the HCL you provide in the testAccAuthenticationPoliciesSettings method. You need to provide HCL with no attributes set. The log settings test does this with some optional HCL it inserts based on a boolean.

File: internal/acctest/authenticationpoliciessettings/authentication_policies_settings_test.go [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+}
+
+data "pingfederate_authentication_policies_settings" "%[1]s" {
+  depends_on = [pingfederate_authentication_policies_settings.%[1]s]
+}`, resourceName,
+		optionalHcl,
+	)
+}
+
+// func testAccAuthenticationPoliciesSettings(resourceName string, resourceModel authenticationPoliciesSettingsResourceModel) string {
```
Comment (henryrecker-pingidentity) [inline internal/acctest/authenticationpoliciessettings/authentication_policies_settings_test.go:71]: You can remove these comments

**Review by henryrecker-pingidentity — APPROVED** [review]
